File: src/test/java/ru/jerael/booktracker/backend/application/usecase/book/DeleteCoverUseCaseImplTest.java
package ru.jerael.booktracker.backend.application.usecase.book;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import ru.jerael.booktracker.backend.domain.exception.NotFoundException;
import ru.jerael.booktracker.backend.domain.model.book.Book;
import ru.jerael.booktracker.backend.domain.model.book.BookStatus;
import ru.jerael.booktracker.backend.domain.repository.BookRepository;
import ru.jerael.booktracker.backend.domain.storage.BookCoverStorage;
import java.time.Instant;
import java.util.Collections;
import java.util.Optional;
import java.util.UUID;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class DeleteCoverUseCaseImplTest {

    @Mock
    private BookRepository bookRepository;

    @Mock
    private BookCoverStorage bookCoverStorage;

    @InjectMocks
    private DeleteCoverUseCaseImpl useCase;

    private final UUID id = UUID.fromString("ee39af7a-a073-4473-878a-1aae34e98bb7");
    private final UUID userId = UUID.fromString("2c5781ea-1bc2-4561-a83d-26106df2526e");
    private final String title = "title";
    private final String author = "author";
    private final BookStatus status = BookStatus.WANT_TO_READ;
    private final Instant createdAt = Instant.ofEpochMilli(1771249699347L);

    @Test
    void execute_WhenBookDoesNotExists_ShouldThrowNotFoundException() {
        when(bookRepository.findByIdAndUserId(id, userId)).thenReturn(Optional.empty());

        assertThrows(NotFoundException.class, () -> useCase.execute(id, userId));

        verifyNoInteractions(bookCoverStorage);
    }

    @Test
    void execute_WhenCoverExists_ShouldDeleteFromStorageAndSaveBookWithNullCover() {
        String coverFileName = "cover.jpg";
        Book book = new Book(id, title, author, coverFileName, status, createdAt, Collections.emptySet());
        when(bookRepository.findByIdAndUserId(id, userId)).thenReturn(Optional.of(book));

        useCase.execute(id, userId);

        Book updatedBook = new Book(id, title, author, null, status, createdAt, Collections.emptySet());
        verify(bookCoverStorage).delete(coverFileName);
        verify(bookRepository).save(updatedBook, userId);
    }

    @Test
    void execute_WhenCoverDoesNotExists_ShouldExitWithoutUpdate() {
        Book book = new Book(id, title, author, null, status, createdAt, Collections.emptySet());
        when(bookRepository.findByIdAndUserId(id, userId)).thenReturn(Optional.of(book));

        useCase.execute(id, userId);

        verify(bookCoverStorage, never()).delete(any());
        verify(bookRepository, never()).save(any(), eq(userId));
    }
}

File: src/test/kotlin/ru/jerael/booktracker/backend/application/usecase/book/DeleteCoverUseCaseImplTest.kt
package ru.jerael.booktracker.backend.application.usecase.book

import io.mockk.*
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import ru.jerael.booktracker.backend.domain.exception.NotFoundException
import ru.jerael.booktracker.backend.domain.repository.BookRepository
import ru.jerael.booktracker.backend.domain.storage.BookCoverStorage
import ru.jerael.booktracker.backend.factory.book.BookDomainFactory
import java.util.*

@ExtendWith(MockKExtension::class)
class DeleteCoverUseCaseImplTest {
    
    @MockK
    private lateinit var bookRepository: BookRepository
    
    @MockK
    private lateinit var bookCoverStorage: BookCoverStorage
    
    @InjectMockKs
    private lateinit var useCase: DeleteCoverUseCaseImpl
    
    private val bookId: UUID = UUID.randomUUID()
    private val userId: UUID = UUID.randomUUID()
    
    @BeforeEach
    fun setUp() {
        every { bookRepository.save(any(), any()) } returns mockk()
        every { bookCoverStorage.delete(any()) } just Runs
    }
    
    @Test
    fun `when book does not exists, execute should throw NotFoundException`() {
        every { bookRepository.findByIdAndUserId(bookId, userId) } returns Optional.empty()
        
        assertThrows(NotFoundException::class.java) { useCase.execute(bookId, userId) }
        
        verify { bookCoverStorage wasNot called }
    }
    
    @Test
    fun `when cover exists, execute should delete from storage and save book with null cover`() {
        val coverFileName = "cover.jpg"
        val book = BookDomainFactory.createBook(
            id = bookId,
            userId = userId,
            coverFileName = coverFileName
        )
        
        every { bookRepository.findByIdAndUserId(bookId, userId) } returns Optional.of(book)
        
        useCase.execute(bookId, userId)
        
        verify { bookRepository.save(match { it.id == bookId && it.coverFileName == null }, userId) }
        verify { bookCoverStorage.delete(coverFileName) }
    }
    
    @Test
    fun `when cover does not exists, execute should exit without update`() {
        val book = BookDomainFactory.createBook(
            id = bookId,
            userId = userId,
            coverFileName = null
        )
        
        every { bookRepository.findByIdAndUserId(bookId, userId) } returns Optional.of(book)
        
        useCase.execute(bookId, userId)
        
        verify(exactly = 0) { bookRepository.save(any(), userId) }
        verify(exactly = 0) { bookCoverStorage.delete(any()) }
    }
}